### PR TITLE
Backend Refactoring for latency

### PR DESCRIPTION
**Pull request by guy-ro**
TODO:
```
- frontend use of API
```

Huge overhaul of our previous approach. Reorganization of a lot of the backend code.

Result is even the most complicated (in terms of nested references) query takes <5s with 1000s of Requests in the database.

- Removed caching as caching doesn't work well with so many references between documents
- Removed datasources in favour of querying with Mongoose directly
- Removed unused attributes from schema
- Renamed misnamed attributes
- Replaced statuses with Dates (an object has a status if the date is not null) - but I added endpoints to return the boolean value expected for each status
- Added denormalization through embedding of partial Request documents in RequestTypes (and made it so that we can do the same for RequestGroups with RequestTypes if needed)
- Added base filtering resolver for all Requests/RequestTypes/RequestGroups/Clients
- Updates take less parameters - some updates must be done through other API endpoints (e.g. to delete a Request, use deleteRequest) in order to keep API simple to understand and use. This means that the API user must call API endpoints sequentially (using onCompleted for Apollo or Promises)
- Added a few endpoints that will likely be useful in the future (e.g. nextRequest returns the next Request within a RequestType/RequestGroup, sorting by creation date)
- Add "...Pages" endpoint for RequestTypes and RequestGroups that returns a page of documents given a skip amount (page offset) and limit amount (page size) in sorted order

```
**DO NOT MERGE!!!
MUST MAKE CHANGES ON HOW FRONTEND USES API FIRST!**
```

### CODE REVIEW

File: server/api/resolvers/clientResolvers.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            Client.findById(_id).exec()
+        }) 
+
+    },
+    clients: async (_, __, { authenticateUser }): Promise<Array<ClientInterface>> => {
+        return authenticateUser().then(() => {
+            Client.find().exec()
+        })
+    },
+    clientsFilter: async (_, { filter, options }, { authenticateUser }): Promise<Array<ClientInterface>> => {
```
Comment (meganniu) [inline server/api/resolvers/clientResolvers.ts]: What does `options` do? LOL i think im missing something

File: server/api/resolvers/clientResolvers.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import { Client, ClientInterface } from '../../database/models/clientModel'
+
+const clientQueryResolvers = {
+    client: async (_, { _id }, { authenticateUser }): Promise<ClientInterface> => {
+        return authenticateUser().then(() => {
+            Client.findById(_id).exec()
```
Comment (meganniu) [inline server/api/resolvers/clientResolvers.ts]: might need to return result of db query
```
return Client.findById(_id).exec()
```

Same suggestion for the other client query resolvers

File: server/api/resolvers/clientResolvers.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    clientsFilter: async (_, { filter, options }, { authenticateUser }): Promise<Array<ClientInterface>> => {
+        return authenticateUser().then(() => {
+            Client.find({ fullName: filter.fullName }).exec()
+        })
+    },
+}
+
+const clientMutationResolvers = {
+    createClient: async (_, { client }, { authenticateUser }): Promise<ClientInterface> => {
+        return authenticateUser().then(async () => {
```
Comment (meganniu) [inline server/api/resolvers/clientResolvers.ts]: Not sure if we need the `async` function declaration. 

Same comment for the other client mutaiton resolvers.

File: server/api/resolvers/requestResolvers.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+const requestQueryResolvers = {
+    request: async (_, { _id }, ___): Promise<RequestInterface> => {
+        return Request.findById(_id).exec()
+    },
+    requests: async (_, __, ___): Promise<Array<RequestInterface>> => {
+        return Request.find().exec()
+    },
+    requestsFilter: async (_, { filter, options }, ___): Promise<Array<RequestInterface>> => {
+        return Request.find().exec()
```
Comment (meganniu) [inline server/api/resolvers/requestResolvers.ts:56]: Are we using the `filter` arg in a future PR?

File: server/api/resolvers/requestResolvers.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    },
+    fulfillRequest: async (_, { _id }, { authenticateUser }): Promise<RequestInterface> => {
+        return authenticateUser().then(async () => {
+            const request = await Request.findById(_id)
+            request.fulfilledAt = new Date()
+            return request.save()
+        })
+    },
+    changeRequestTypeForRequest: async (_, { request, requestType }, { authenticateUser }): Promise<RequestInterface> => {
+        return authenticateUser().then(async () => {
```
Comment (meganniu) [inline server/api/resolvers/requestResolvers.ts:130]: might not need the `async` functino declaration

File: server/scripts/seed.ts [hunk 6]
```diff
@@ -37,72 +37,64 @@ const randomDate = (start = startDate, end = endDate) => {
...
-  const dateFulfilled = randomDate(new Date(dateCreated))
-  const dateUpdated = isFulfilled ? randomDate(new Date(dateFulfilled)) : randomDate(new Date(dateCreated))
 
   const request = new Request({
-    _id: requestID,
-    requestId: faker.random.alphaNumeric(6),
-    client: faker.random.arrayElement(clientIDs),
-    deleted: isDeleted,
+    _id: mongoose.Types.ObjectId(),
+    quantity: Math.floor(Math.random() * 15) + 1,
```
Comment (meganniu) [inline server/scripts/seed.ts]: Can we store `15` in a variable?
```
maxQuantityPerRequest = 15
```

File: server/scripts/seed.ts [hunk 7]
```diff
@@ -37,72 +37,64 @@ const randomDate = (start = startDate, end = endDate) => {
...
     fulfilled: isFulfilled,
```
Comment (meganniu) [inline server/scripts/seed.ts]: Not sure that `fulfilled` is part of the new `RequestInterface`.
Do we still need to include it in the seeder?

File: server/scripts/seed.ts [hunk 8]
```diff
@@ -37,72 +37,64 @@ const randomDate = (start = startDate, end = endDate) => {
...
-  return createSavePromise(type, errMsg)
 }
 
-const createRequestGroup = (groupID, groupName, typeIDs, errMsg) => {
-  const group = new RequestGroup({
-    _id: groupID,
-    name: groupName,
+// create RequestGroup model object without references
+const createRequestGroup = () => {
+  const dateCreated = randomDate()
```
Comment (meganniu) [inline server/scripts/seed.ts]: The way `dateCreated` is initialized here is different from how dates are initialized in `createRequest`

Do we need to do it like this instead?
```
const dateCreated = new Date(randomDate())
```

File: server/server.ts [hunk 9]
```diff
@@ -44,12 +33,7 @@ const corsPolicy = {
 // -----------------------------------------------------------------------------
 
 // connect to MongoDB and setup data sources
-connectDB(() => {
-  ClientCache.init();
-  RequestCache.init();
-  RequestTypeCache.init();
-  RequestGroupCache.init();
-});
+connectDB(() => {});
```
Comment (meganniu) [inline server/server.ts]: Nit: replace with `connectDB()`

File: server/server.ts [hunk 10]
```diff
@@ -123,17 +107,11 @@ async function gqlServer() {
 
         const user = await getUser(req.cookies.session);
         if (!user || !user.id)
-          throw new AuthenticationError("Authentication Error");
+          throw new ForbiddenError("Authentication Error");
```
Comment (meganniu) [inline server/server.ts]: Nit: replace error string with "Forbidden Error"

File: server/api/resolvers/requestResolvers.ts [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        })
+    },
+    deleteRequest: async (_, { _id }, { authenticateUser }): Promise<RequestInterface> => {
+        return authenticateUser().then(async () => {
+            const request = await Request.findById(_id)
+            request.deletedAt = new Date()
+            return request.save()
+        })
+    },
+    fulfillRequest: async (_, { _id }, { authenticateUser }): Promise<RequestInterface> => {
```
Comment (meganniu) [inline server/api/resolvers/requestResolvers.ts:105]: We're going to need an endpoint to unfulfill a request when integrating with the frontend.

The new endpoint doesn't need to be included in this PR tho. I just wanted to make a note

File: server/database/models/newSchema.txt [hunk 12]
```diff
@@ -0,0 +1,56 @@
+interface RequestGroupInterface {
```
Comment (meganniu) [inline server/database/models/newSchema.txt]: LOL is this file supposed to be checked in?

**Review by meganniu — CHANGES_REQUESTED** [review]
Looking good @guy-romm! Thanks for taking care of all this refactoring work

Left a few questions in the first pass, I'll do some more thorough testing

File: server/api/resolvers/clientResolvers.ts (hunk 1 above)
Comment (guy-ro) [reply server/api/resolvers/clientResolvers.ts]: Implemented as a proof of concept, will comment it out for future developers to see as an example of how to implement filtering

File: server/api/resolvers/clientResolvers.ts (hunk 2 above)
Comment (guy-ro) [reply server/api/resolvers/clientResolvers.ts]: Good catch

File: server/api/resolvers/clientResolvers.ts (hunk 3 above)
Comment (guy-ro) [reply server/api/resolvers/clientResolvers.ts]: Needed in order to use authenticateUser with async 'then'

File: server/api/resolvers/requestResolvers.ts (hunk 4 above)
Comment (guy-ro) [reply server/api/resolvers/requestResolvers.ts:56]: See another comment

File: server/api/resolvers/requestResolvers.ts (hunk 5 above)
Comment (guy-ro) [reply server/api/resolvers/requestResolvers.ts:130]: See another comment

File: server/scripts/seed.ts (hunk 6 above)
Comment (guy-ro) [reply server/scripts/seed.ts]: Done

File: server/scripts/seed.ts (hunk 7 above)
Comment (guy-ro) [reply server/scripts/seed.ts]: Fixed

File: server/scripts/seed.ts (hunk 8 above)
Comment (guy-ro) [reply server/scripts/seed.ts]: Modified for consistency, but this didn't matter as Mongoose takes millisecond from unix epoch and Date objects just the same.

File: server/server.ts (hunk 9 above)
Comment (guy-ro) [reply server/server.ts]: Done

File: server/server.ts (hunk 10 above)
Comment (guy-ro) [reply server/server.ts]: Done

File: server/database/models/newSchema.txt (hunk 12 above)
Comment (guy-ro) [reply server/database/models/newSchema.txt]: Leaving this for now, just stored some ideas for what subdocuments could look like. Will remove when sorting/pagination is done.

File: server/api/resolvers/requestResolvers.ts (hunk 11 above)
Comment (guy-ro) [reply server/api/resolvers/requestResolvers.ts:105]: Implemented lol

File: server/api/resolvers/clientResolvers.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+            const newClient = new Client({...client})
+            return newClient.save()
+        })
+    },
+    updateClient: async (_, { client }, { authenticateUser }): Promise<ClientInterface> => {
+        return authenticateUser().then(async () => {
+            const currentClient = await Client.findById(client._id)
```
Comment (kevin-wangg) [inline server/api/resolvers/clientResolvers.ts]: Should we verify the client object contains the field _id? Same with the other endpoints.

File: server/api/resolvers/requestResolvers.ts [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        return authenticateUser().then(async () => {
+            const request = await Request.findById(_id)
+            request.fulfilledAt = undefined
+            return request.save()
+        })
+    },
+    changeRequestTypeForRequest: async (_, { request, requestType }, { authenticateUser }): Promise<RequestInterface> => {
+        return authenticateUser().then(async () => {
+            return sessionHandler(async (session) => {
+                const modifiedRequestObject = await Request.findById(request).session(session)
```
Comment (kevin-wangg) [inline server/api/resolvers/requestResolvers.ts]: why are we passing request into the findById. should it be request._id?

File: server/api/resolvers/requestGroupResolvers.ts [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+    if (sortedNextRequests.length === 0) {
+        return null
+    }
+
+    return sortedNextRequests[0]
+}
+
+const requestGroupQueryResolvers = {
+    requestGroup: async (_, { _id }, ___): Promise<RequestGroupInterface> => {
+        return RequestGroup.findById(_id).exec()
```
Comment (Anand-Issac) [inline server/api/resolvers/requestGroupResolvers.ts]: would you have to await here to get response back from RequestGroup.findById

File: server/api/resolvers/requestGroupResolvers.ts [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+    },
+    requestGroups: async (_, __, ___): Promise<Array<RequestGroupInterface>> => {
+        return RequestGroup.find().exec()
+    },
+    requestGroupsFilter: async (_, { filter, options }, ___): Promise<Array<RequestGroupInterface>> => {
+        return RequestGroup.find().exec()
```
Comment (Anand-Issac) [inline server/api/resolvers/requestGroupResolvers.ts:68]: shouldn't the filter parameter be used in requestGroupsFilter

File: server/api/resolvers/requestGroupResolvers.ts [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+            requestGroup.deletedAt = new Date()
+            return requestGroup.save()
+        })
+    },
+}
+
+const requestGroupResolvers = {
+    requestTypes: async (parent, __, ___): Promise<Array<RequestTypeInterface>> => {
+        return parent.requestTypes.map((requestTypeEmbedding) => {
+            return RequestType.findById(requestTypeEmbedding._id)
```
Comment (Anand-Issac) [inline server/api/resolvers/requestGroupResolvers.ts]: await needed here?

File: server/api/resolvers/requestGroupResolvers.ts [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+        })
+    },
+    deleted: (parent, __, ___): boolean => {
+        return parent.deletedAt !== undefined
+    },
+    countOpenRequests: async (parent, __, ___): Promise<number> => {
+        const countOpenRequestsPerType: Array<number> = await Promise.all(parent.requestTypes.map(async (requestTypeEmbedding) => {
```
Comment (Anand-Issac) [inline server/api/resolvers/requestGroupResolvers.ts]: why do some resolvers require authenticateUser() but others dont?

File: server/api/resolvers/requestGroupResolvers.ts [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+
+                if (nextRequestEmbedding === null) {
+                    return null
+                }
+
+                return Request.findById(nextRequestEmbedding._id).session(session)
+            })
+        })
+    },
+    nextRecipient: async (parent, __, { authenticateUser }): Promise<ClientInterface> => {
```
Comment (Anand-Issac) [inline server/api/resolvers/requestGroupResolvers.ts]: sorry but what does nextRecipient mean for a requestGroup?

File: server/api/resolvers/requestResolvers.ts [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+const requestMutationResolvers = {
+    createRequest: async (_, { request }, { authenticateUser }): Promise<RequestInterface> => {
+        return authenticateUser().then(async () => { 
+            return sessionHandler(async (session) => {
+                const newRequestObject = new Request({...request})
+                const newRequest = await newRequestObject.save({ session: session })
+                
+                const requestType = await RequestType.findById(newRequest.requestType).session(session)
```
Comment (Anand-Issac) [inline server/api/resolvers/requestResolvers.ts]: should this be RequestType.findById(newRequest.requestType._id) ?

File: server/api/resolvers/requestTypeResolvers.ts [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+            return Request.findById(requestEmbedding._id)
+        })
+    },
+    deleted: (parent, __, ___): boolean => {
+        return parent.deletedAt !== undefined
+    },
+    openRequests: async (parent, __, ___): Promise<Array<RequestInterface>> => {
+        return filterOpenRequestEmbeddings(parent.requests)
+            .map((requestEmbedding) => {
+                return Request.findById(requestEmbedding._id)
```
Comment (Anand-Issac) [inline server/api/resolvers/requestTypeResolvers.ts]: await Request.findById

File: server/api/resolvers/requestTypeResolvers.ts [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+    },
+}
+
+const requestTypeResolvers = {
+    requestGroup: async (parent, __, ___): Promise<RequestGroupInterface> => {
+        return RequestGroup.findById(parent.requestGroup)
+    },
+    requests: async (parent, __, ___): Promise<Array<RequestInterface>> => {
+        return parent.requests.map((requestEmbedding) => {
+            return Request.findById(requestEmbedding._id)
```
Comment (Anand-Issac) [inline server/api/resolvers/requestTypeResolvers.ts]: await Request.findById?

File: server/api/resolvers/clientResolvers.ts (hunk 13 above)
Comment (guy-ro) [reply server/api/resolvers/clientResolvers.ts]: In the graphql schema, _id is set as mandatory

File: server/api/resolvers/requestResolvers.ts (hunk 14 above)
Comment (guy-ro) [reply server/api/resolvers/requestResolvers.ts]: request's type is ID. Should the name be changed?

Comment (guy-ro) [reply server/api/resolvers/requestResolvers.ts]: Modified it to "requestId" and "requestTypeId"

File: server/api/resolvers/requestGroupResolvers.ts (hunk 15 above)
Comment (guy-ro) [reply server/api/resolvers/requestGroupResolvers.ts]: No. GraphQL resolvers handle awaiting Promises before returning the API response

File: server/api/resolvers/requestGroupResolvers.ts (hunk 16 above)
Comment (guy-ro) [reply server/api/resolvers/requestGroupResolvers.ts:68]: Yes, but it isn't fully implemented. Commented out for now.

File: server/api/resolvers/requestGroupResolvers.ts (hunk 17 above)
Comment (guy-ro) [reply server/api/resolvers/requestGroupResolvers.ts]: Addressed in other comment

File: server/api/resolvers/requestGroupResolvers.ts (hunk 18 above)
Comment (guy-ro) [reply server/api/resolvers/requestGroupResolvers.ts]: Depends on whether they are accessing Client information. We have restricted access to that portion of the API to authenticated users only

File: server/api/resolvers/requestGroupResolvers.ts (hunk 19 above)
Comment (guy-ro) [reply server/api/resolvers/requestGroupResolvers.ts]: The Client whose Request is next in line (FIFO)

File: server/api/resolvers/requestResolvers.ts (hunk 20 above)
Comment (guy-ro) [reply server/api/resolvers/requestResolvers.ts]: Mongoose model shows Request's requestType is an ID

Comment (guy-ro) [reply server/api/resolvers/requestResolvers.ts]: So newRequest.requestType is the ID

File: server/api/resolvers/requestTypeResolvers.ts (hunk 22 above)
Comment (guy-ro) [reply server/api/resolvers/requestTypeResolvers.ts]: Addressed in previous comment

File: server/api/resolvers/requestTypeResolvers.ts (hunk 21 above)
Comment (guy-ro) [reply server/api/resolvers/requestTypeResolvers.ts]: GraphQL awaits this just like the previous comments

File: server/api/resolvers/requestGroupResolvers.ts (hunk 18 above)
Comment (meganniu) [reply server/api/resolvers/requestGroupResolvers.ts]: yup, authenticateUser() is used to restrict usage of mutation endpoints and endpoints that expose client info

**Review by meganniu — APPROVED** [review]
LGTM!!! Thanks again for all your hard work

**Review by kevin-wangg — APPROVED** [review]
Thanks Guy! LGTM!

**Review by Anand-Issac — APPROVED** [review]
LGTM

**Pull request merged**